File: apps/ios/AgentDeck/UI/DiffView.swift
import SwiftUI

/// What the session changed on disk, one file at a time.
/// Mirrored from `DiffView` / `DiffFileCard` on Android.
struct DiffView: View {
    var files: [AgentFileChange]
    /// Whether the fetch has come back at all. Claiming "0 files changed" while
    /// it is still in flight states a result the screen does not have yet.
    var loaded: Bool

    @State private var allExpanded = true
    /// Bumped by the expand-all control. Each card keys its own state on this,
    /// so one press moves every card without the parent owning per-card state.
    @State private var expandRevision = 0

    private var additions: Int { files.reduce(0) { $0 + $1.additions } }
    private var deletions: Int { files.reduce(0) { $0 + $1.deletions } }

    var body: some View {
        VStack(spacing: 0) {
            header
            if files.isEmpty {
                empty
            } else {
                ScrollView {
                    LazyVStack(spacing: 12) {
                        ForEach(files) { file in
                            DiffFileCard(file: file, expandedByDefault: allExpanded, expandRevision: expandRevision)
                        }
                    }
                    .padding(.horizontal, 12)
                    .padding(.vertical, 12)
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Palette.ink)
    }

    private var header: some View {
        HStack(spacing: 8) {
            Image(systemName: "plusminus")
                .font(.system(size: 14))
                .foregroundStyle(Palette.muted)
            Text(!loaded && files.isEmpty ? "Changes" : "\(files.count) \(files.count == 1 ? "file" : "files") changed")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(Palette.text.opacity(0.86))
            Spacer(minLength: 0)
            if additions > 0 {
                Text("+\(additions)")
                    .font(.system(size: 11, weight: .semibold, design: .monospaced))
                    .foregroundStyle(Palette.signal)
            }
            if deletions > 0 {
                Text("\u{2212}\(deletions)")
                    .font(.system(size: 11, weight: .semibold, design: .monospaced))
                    .foregroundStyle(Palette.danger)
            }
            if !files.isEmpty {
                Button {
                    allExpanded.toggle()
                    expandRevision += 1
                } label: {
                    Image(systemName: allExpanded ? "arrow.down.right.and.arrow.up.left" : "arrow.up.left.and.arrow.down.right")
                        .font(.system(size: 14))
                        .foregroundStyle(Palette.muted)
                        .frame(width: 40, height: 40)
                }
                .accessibilityLabel(allExpanded ? "Collapse all files" : "Expand all files")
            }
        }
        .padding(.leading, 16)
        .padding(.trailing, 6)
        .frame(height: 48)
        .background(Color(hex: 0x0C1014))
    }

    private var empty: some View {
        VStack(spacing: 6) {
            Image(systemName: "plusminus")
                .font(.system(size: 28))
                .foregroundStyle(Palette.muted)
                .padding(.bottom, 6)
            // Until the session's changes have been fetched, "none" is not
            // yet known.
            Text(loaded ? "No captured changes" : "Loading changes…")
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(Palette.text)
            if loaded {
                Text("Edits and writes exposed by this runtime will appear here.")
                    .font(.system(size: 13))
                    .foregroundStyle(Palette.muted)
                    .multilineTextAlignment(.center)
            }
        }
        .padding(.horizontal, 28)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }
}

/// Long file rewrites arrive as thousands of `+` lines; render a readable slice
/// until asked for the rest.
private let diffLineBudget = 300

private struct DiffFileCard: View {
    var file: AgentFileChange
    var expandedByDefault: Bool
    var expandRevision: Int

    @State private var expanded = true
    @State private var showAllLines = false

    private var truncated: Bool { file.lineCount > diffLineBudget && !showAllLines }

    var body: some View {
        VStack(spacing: 0) {
            head
            if expanded {
                Divider().overlay(Palette.line)
                body(for: file)
                if file.lineCount > diffLineBudget {
                    Divider().overlay(Palette.line)
                    Button {
                        showAllLines.toggle()
                    } label: {
                        Text(truncated ? "Show all \(file.lineCount) lines" : "Show first \(diffLineBudget) lines")
                            .font(.system(size: 12))
                            .foregroundStyle(Palette.blue)
                            .frame(maxWidth: .infinity, minHeight: 44)
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .background(RoundedRectangle(cornerRadius: 14).fill(Palette.surface))
        .overlay(RoundedRectangle(cornerRadius: 14).stroke(Palette.line, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .onAppear { expanded = expandedByDefault }
        .onChange(of: expandRevision) { _, _ in expanded = expandedByDefault }
    }

    private var head: some View {
        Button { expanded.toggle() } label: {
            HStack(spacing: 8) {
                Image(systemName: "doc.text")
                    .font(.system(size: 14))
                    .foregroundStyle(Palette.blue)
                // The leading directories are the droppable part; the file name
                // is what a person is reading.
                Text(file.path)
                    .font(.system(size: 11, design: .monospaced))
                    .foregroundStyle(Palette.text.opacity(0.9))
                    .lineLimit(1)
                    .truncationMode(.head)
                    .frame(maxWidth: .infinity, alignment: .leading)
                if file.additions > 0 {
                    Text("+\(file.additions)")
                        .font(.system(size: 10, design: .monospaced))
                        .foregroundStyle(Palette.signal)
                }
                if file.deletions > 0 {
                    Text("\u{2212}\(file.deletions)")
                        .font(.system(size: 10, design: .monospaced))
                        .foregroundStyle(Palette.danger)
                }
                Image(systemName: expanded ? "chevron.up" : "chevron.down")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Palette.muted)
            }
            .padding(.leading, 12)
            .padding(.trailing, 10)
            .padding(.vertical, 11)
        }
        .buttonStyle(.plain)
    }

    /// One horizontal scroller for the whole file, so every line's tint runs to
    /// the same edge. Scrolling each row separately would let two lines of the
    /// same hunk sit at different offsets.
    private func body(for file: AgentFileChange) -> some View {
        ScrollView(.horizontal, showsIndicators: false) {
            VStack(alignment: .leading, spacing: 0) {
                ForEach(rows) { row in
                    switch row.content {
                    case .separator(let label):
                        Text(label)
                            .font(.system(size: 10, design: .monospaced))
                            .foregroundStyle(Palette.blue)
                            .padding(.horizontal, 10)
                            .padding(.vertical, 5)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Palette.blue.opacity(0.06))
                    case .line(let line):
                        DiffLineRow(line: line, showLineNumbers: file.hasLineNumbers)
                    }
                }
            }
            // A horizontal scroller hands children unbounded width, so a row
            // fills only its own text and each line's tint stops where its
            // characters do. Pinning the column to the viewport gives the rows
            // a real width to fill, edge to edge.
            .frame(minWidth: horizontalViewportWidth, alignment: .leading)
        }
        .textSelection(.enabled)
    }

    /// Enough that a short diff still tints the full card. The scroller grows
    /// past it for anything wider.
    private var horizontalViewportWidth: CGFloat {
        UIScreen.main.bounds.width - 48
    }

    /// The card's hunks flattened into one list, cut at the budget.
    ///
    /// Flattened before rendering rather than counted inside the view builder:
    /// a running total cannot be kept across a `ForEach`, and a per-hunk slice
    /// would let a file of five small hunks render five budgets' worth.
    private var rows: [DiffRow] {
        var result: [DiffRow] = []
        let budget = truncated ? diffLineBudget : Int.max
        for (index, hunk) in file.hunks.enumerated() {
            if result.count >= budget { break }
            if index > 0 {
                result.append(DiffRow(
                    id: "sep:\(hunk.id)",
                    content: .separator("Change \(index + 1) \u{00B7} \(messageClock(hunk.createdAt))")
                ))
            }
            for line in hunk.lines {
                if result.count >= budget { break }
                result.append(DiffRow(id: "\(hunk.id):\(line.index)", content: .line(line)))
            }
        }
        return result
    }
}

private struct DiffRow: Identifiable {
    enum Content {
        /// Where one runtime edit ends and the next begins, in a file both
        /// touched.
        case separator(String)
        case line(AgentDiffLine)
    }

    var id: String
    var content: Content
}

private struct DiffLineRow: View {
    var line: AgentDiffLine
    var showLineNumbers: Bool

    var body: some View {
        HStack(alignment: .top, spacing: 0) {
            if showLineNumbers {
                Text(number.map(String.init) ?? "")
                    .font(.system(size: 10, design: .monospaced))
                    .foregroundStyle(Palette.muted.opacity(0.55))
                    .frame(width: 40, alignment: .trailing)
                    .padding(.trailing, 6)
            }
            Text(marker)
                .font(.system(size: 11, design: .monospaced))
                .foregroundStyle(foreground)
                .frame(width: 24)
                .frame(maxHeight: .infinity)
                .background(foreground.opacity(0.08))
            Text(text)
                .font(.system(size: 11, design: .monospaced))
                .foregroundStyle(foreground)
                .lineLimit(1)
                .fixedSize(horizontal: true, vertical: false)
                .padding(.horizontal, 8)
                .padding(.vertical, 2)
            Spacer(minLength: 0)
        }
        .frame(minHeight: 24)
        .background(background)
    }

    private var text: String {
        if line.kind == .header { return hunkHeaderContext(line.text) ?? line.text }
        return line.text.isEmpty ? " " : line.text
    }

    /// Deletions are positioned in the old file, everything else in the new one.
    private var number: Int? { line.kind == .deletion ? line.oldLine : line.newLine }

    private var marker: String {
        switch line.kind {
        case .addition: "+"
        case .deletion: "\u{2212}"
        default: " "
        }
    }

    private var foreground: Color {
        switch line.kind {
        case .addition: Palette.signal
        case .deletion: Palette.danger
        case .header: Palette.blue
        case .context: Palette.text.opacity(0.78)
        }
    }

    private var background: Color {
        switch line.kind {
        case .addition: Palette.signal.opacity(0.10)
        case .deletion: Palette.danger.opacity(0.10)
        case .header: Palette.blue.opacity(0.08)
        case .context: .clear
        }
    }
}

/// A conversation timestamps to the minute.
func messageClock(_ value: String) -> String {
    guard let date = Timestamps.parse(value) else { return "" }
    return date.formatted(date: .omitted, time: .shortened)
}

File: apps/android/mobile/src/main/java/dev/agentdeck/mobile/DiffView.kt
package dev.agentdeck.mobile

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.agentdeck.shared.*

/** Every file the session touched, as a sheet over the conversation. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ChangesSheet(files: List<AgentFileChange>, loaded: Boolean, onDismiss: () -> Unit) {
    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = Surface) {
        DiffView(files, loaded, Modifier.fillMaxWidth().heightIn(max = 560.dp))
    }
}

@Composable
internal fun DiffView(files: List<AgentFileChange>, loaded: Boolean, modifier: Modifier = Modifier) {
    val additions = files.sumOf { it.additions }
    val deletions = files.sumOf { it.deletions }
    var allExpanded by rememberSaveable { mutableStateOf(true) }
    var expandRevision by rememberSaveable { mutableStateOf(0) }
    Column(modifier.fillMaxWidth().background(Ink)) {
        Row(
            modifier = Modifier.fillMaxWidth().height(48.dp).background(Color(0xFF0C1014)).padding(start = 16.dp, end = 6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(Icons.Rounded.Difference, null, tint = Muted, modifier = Modifier.size(17.dp))
            Spacer(Modifier.width(8.dp))
            // Claiming "0 files changed" while the fetch is still in flight
            // states a result the screen does not have yet. The body says it is
            // loading; this bar just holds the count until there is one.
            Text(
                if (!loaded && files.isEmpty()) "Changes" else "${plural(files.size, "file")} changed",
                color = Text.copy(alpha = 0.86f),
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
            )
            Spacer(Modifier.weight(1f))
            if (additions > 0) Text("+$additions", color = Signal, fontFamily = FontFamily.Monospace, fontSize = 11.sp, fontWeight = FontWeight.SemiBold)
            if (additions > 0 && deletions > 0) Spacer(Modifier.width(10.dp))
            if (deletions > 0) Text("−$deletions", color = Danger, fontFamily = FontFamily.Monospace, fontSize = 11.sp, fontWeight = FontWeight.SemiBold)
            if (files.isNotEmpty()) {
                Spacer(Modifier.width(4.dp))
                IconButton(
                    onClick = {
                        allExpanded = !allExpanded
                        expandRevision += 1
                    },
                    modifier = Modifier.size(40.dp),
                ) {
                    Icon(
                        if (allExpanded) Icons.Rounded.UnfoldLess else Icons.Rounded.UnfoldMore,
                        if (allExpanded) "Collapse all files" else "Expand all files",
                        tint = Muted,
                        modifier = Modifier.size(19.dp),
                    )
                }
            }
        }
        if (files.isEmpty()) {
            Column(
                Modifier.fillMaxSize().padding(horizontal = 28.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Center,
            ) {
                Icon(Icons.Rounded.Difference, null, tint = Muted, modifier = Modifier.size(32.dp))
                Spacer(Modifier.height(12.dp))
                // Until the session's changes have been fetched, "none" is not yet known.
                Text(if (loaded) "No captured changes" else "Loading changes…", fontWeight = FontWeight.SemiBold)
                if (loaded) {
                    Spacer(Modifier.height(6.dp))
                    Text("Edits and writes exposed by this runtime will appear here.", color = Muted, fontSize = 13.sp, lineHeight = 19.sp, textAlign = TextAlign.Center)
                }
            }
        } else {
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(horizontal = 12.dp, vertical = 12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                items(files, key = { "diff:${it.path}" }) { file ->
                    DiffFileCard(file, expandedByDefault = allExpanded, expandRevision = expandRevision)
                }
            }
        }
    }
}

@Composable
internal fun DiffFileCard(file: AgentFileChange, expandedByDefault: Boolean, expandRevision: Int) {
    var expanded by rememberSaveable(file.path, expandRevision) { mutableStateOf(expandedByDefault) }
    var showAllLines by rememberSaveable(file.path) { mutableStateOf(false) }
    val truncated = file.lineCount > DIFF_LINE_BUDGET && !showAllLines
    Surface(
        shape = RoundedCornerShape(14.dp),
        color = Surface,
        border = BorderStroke(1.dp, Line),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Column {
            Row(
                modifier = Modifier.fillMaxWidth().clickable { expanded = !expanded }.padding(start = 12.dp, end = 8.dp, top = 11.dp, bottom = 11.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(Icons.Rounded.Description, null, tint = Blue, modifier = Modifier.size(17.dp))
                Spacer(Modifier.width(8.dp))
                // Start-ellipsis keeps the file name readable; the leading directories are the droppable part.
                Text(file.path, color = Text.copy(alpha = 0.9f), fontFamily = FontFamily.Monospace, fontSize = 11.sp, maxLines = 1, overflow = TextOverflow.StartEllipsis, modifier = Modifier.weight(1f))
                if (file.additions > 0) Text("+${file.additions}", color = Signal, fontFamily = FontFamily.Monospace, fontSize = 10.sp)
                if (file.additions > 0 && file.deletions > 0) Spacer(Modifier.width(8.dp))
                if (file.deletions > 0) Text("−${file.deletions}", color = Danger, fontFamily = FontFamily.Monospace, fontSize = 10.sp)
                Spacer(Modifier.width(5.dp))
                Icon(if (expanded) Icons.Rounded.ExpandLess else Icons.Rounded.ExpandMore, if (expanded) "Collapse file" else "Expand file", tint = Muted, modifier = Modifier.size(18.dp))
            }
            if (expanded) {
                HorizontalDivider(color = Line)
                SelectionContainer {
                    BoxWithConstraints(Modifier.fillMaxWidth()) {
                        val viewportWidth = maxWidth
                        Box(Modifier.fillMaxWidth().horizontalScroll(rememberScrollState())) {
                            // A horizontal scroller hands children infinite width, so fillMaxWidth() on a
                            // row is a no-op and each line's tint stops at its own text. Pinning the column
                            // to its widest line gives the rows a real width to fill, edge to edge.
                            Column(Modifier.widthIn(min = viewportWidth).width(IntrinsicSize.Max)) {
                                var rendered = 0
                                val budget = if (truncated) DIFF_LINE_BUDGET else Int.MAX_VALUE
                                file.hunks.forEachIndexed { index, hunk ->
                                    if (rendered >= budget) return@forEachIndexed
                                    if (index > 0) {
                                        Row(Modifier.fillMaxWidth().background(Blue.copy(alpha = 0.06f)).padding(horizontal = 10.dp, vertical = 5.dp)) {
                                            Text("Change ${index + 1} · ${formatMessageTime(hunk.createdAt)}", color = Blue, fontFamily = FontFamily.Monospace, fontSize = 10.sp)
                                        }
                                    }
                                    for (line in hunk.lines) {
                                        if (rendered >= budget) break
                                        DiffLineRow(line, showLineNumbers = file.hasLineNumbers)
                                        rendered += 1
                                    }
                                }
                            }
                        }
                    }
                }
                if (file.lineCount > DIFF_LINE_BUDGET) {
                    HorizontalDivider(color = Line)
                    TextButton(
                        onClick = { showAllLines = !showAllLines },
                        modifier = Modifier.fillMaxWidth().heightIn(min = 44.dp),
                    ) {
                        Text(
                            if (truncated) "Show all ${file.lineCount} lines" else "Show first $DIFF_LINE_BUDGET lines",
                            color = Blue,
                            fontSize = 12.sp,
                        )
                    }
                }
            }
        }
    }
}

@Composable
internal fun DiffLineRow(line: AgentDiffLine, showLineNumbers: Boolean) {
    val background = when (line.kind) {
        DiffLineKind.Addition -> Signal.copy(alpha = 0.10f)
        DiffLineKind.Deletion -> Danger.copy(alpha = 0.10f)
        DiffLineKind.Header -> Blue.copy(alpha = 0.08f)
        DiffLineKind.Context -> Color.Transparent
    }
    val foreground = when (line.kind) {
        DiffLineKind.Addition -> Signal
        DiffLineKind.Deletion -> Danger
        DiffLineKind.Header -> Blue
        DiffLineKind.Context -> Text.copy(alpha = 0.78f)
    }
    val marker = when (line.kind) {
        DiffLineKind.Addition -> "+"
        DiffLineKind.Deletion -> "−"
        else -> " "
    }
    // Deletions are positioned in the old file, everything else in the new one.
    val lineNumber = if (line.kind == DiffLineKind.Deletion) line.oldLine else line.newLine
    Row(Modifier.fillMaxWidth().background(background).heightIn(min = 24.dp), verticalAlignment = Alignment.Top) {
        if (showLineNumbers) {
            Box(
                Modifier.width(40.dp).heightIn(min = 24.dp).padding(end = 6.dp),
                contentAlignment = Alignment.TopEnd,
            ) {
                Text(
                    lineNumber?.toString().orEmpty(),
                    color = Muted.copy(alpha = 0.55f),
                    fontFamily = FontFamily.Monospace,
                    fontSize = 10.sp,
                    lineHeight = 20.sp,
                )
            }
        }
        Box(Modifier.width(24.dp).heightIn(min = 24.dp).background(foreground.copy(alpha = 0.08f)), contentAlignment = Alignment.TopCenter) {
            Text(marker, color = foreground, fontFamily = FontFamily.Monospace, fontSize = 11.sp, lineHeight = 20.sp)
        }
        Text(
            if (line.kind == DiffLineKind.Header) hunkHeaderContext(line.text) ?: line.text else line.text.ifEmpty { " " },
            color = foreground,
            fontFamily = FontFamily.Monospace,
            fontSize = 11.sp,
            lineHeight = 20.sp,
            softWrap = false,
            modifier = Modifier.padding(horizontal = 8.dp, vertical = 2.dp),
        )
    }
}

/** Long file rewrites arrive as thousands of `+` lines; render a readable slice until asked for the rest. */
internal const val DIFF_LINE_BUDGET = 300
